### PR TITLE
API Support public/ webroot folder

### PR DESCRIPTION
**Pull request by tractorcow**
Also Update filesystem utilities for path manipulations

Parent story https://github.com/silverstripe/silverstripe-framework/issues/7419

### CODE REVIEW

File: tests/php/FilesystemTest.php [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+     * @param string $error Expected path
+     */
+    public function testJoinPathsErrors($args, $error)
+    {
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage($error);
+        Filesystem::joinPaths($args);
+    }
+
+    public function providerTestJoinPathsErrors()
```
Comment (dhensby) [inline tests/php/FilesystemTest.php]: only `test` methods should be public, providers should be `protected`

Comment (robbieaverill) [reply tests/php/FilesystemTest.php]: That’s not true in phpunit 5, though I wish it was :-D

Comment (tractorcow) [reply tests/php/FilesystemTest.php]: phpunit docs declare public providers still.

https://phpunit.de/manual/current/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers

Can you show me a reference to protected providers being preferred?

Comment (tractorcow) [reply tests/php/FilesystemTest.php]:
```md
A data provider method must be public and either return an array of arrays or an object that implements the Iterator interface and yields an array for each iteration step. For each array that is part of the collection the test method will be called with the contents of the array as its arguments.
```

Comment (tractorcow) [reply tests/php/FilesystemTest.php]: Same docs in 7.0 of phpunit.

Comment (dhensby) [reply tests/php/FilesystemTest.php]: Yikes. I've always made them protected. Maybe it's a Codeception thing, then... My bad

**Pull request merged**